mlrun/model_monitoring/applications/histogram_data_drift.py

```diff
@@ -264,45 +270,59 @@ def _log_drift_artifacts(
+            sample_set_statistics=self._remove_timestamp_feature(
+                monitoring_context.sample_df_stats
+            ),
+            inputs_statistics=monitoring_context.feature_stats,
             metrics_per_feature=metrics_per_feature,
             drift_per_feature_values=drift_per_feature_values,
+            monitoring_context=monitoring_context,
         )
 
     def do_tracking(
         self,
-        application_name: str,
-        sample_df_stats: mlrun.common.model_monitoring.helpers.FeatureStats,
-        feature_stats: mlrun.common.model_monitoring.helpers.FeatureStats,
-        sample_df: DataFrame,
-        start_infer_time: Timestamp,
-        end_infer_time: Timestamp,
-        latest_request: Timestamp,
-        endpoint_id: str,
-        output_stream_uri: str,
-    ) -> list[ModelMonitoringApplicationResult]:
+        monitoring_context: mm_context.MonitoringApplicationContext,
+    ) -> list[
+        typing.Union[
+            mm_results.ModelMonitoringApplicationResult,
+            mm_results.ModelMonitoringApplicationMetric,
+        ]
+    ]:
         """
         Calculate and return the data drift metrics, averaged over the features.
 
-        Refer to `ModelMonitoringApplicationBase` for the meaning of the
+        Refer to `ModelMonitoringApplicationBaseV2` for the meaning of the
         function arguments.
         """
-        self.context.logger.debug("Starting to run the application")
+        monitoring_context.logger.debug("Starting to run the application")
+        if not monitoring_context.feature_stats:
+            monitoring_context.logger.info(
+                "No feature statistics found, skipping the application. \n"
+                "In order to run the application, training set must be provided when logging the model."
+            )
+            return []
         metrics_per_feature = self._compute_metrics_per_feature(
-            sample_df_stats=self.dict_to_histogram(sample_df_stats),
-            feature_stats=self.dict_to_histogram(feature_stats),
+            monitoring_context=monitoring_context
         )
-        self.context.logger.debug("Saving artifacts")
+        monitoring_context.logger.debug("Saving artifacts")
         self._log_drift_artifacts(
-            inputs_statistics=feature_stats,
-            sample_set_statistics=sample_df_stats,
+            monitoring_context=monitoring_context,
             metrics_per_feature=metrics_per_feature,
         )
-        self.context.logger.debug("Computing average per metric")
-        results = self._get_results(metrics_per_feature)
-        self.context.logger.debug("Finished running the application", results=results)
-        return results
+        monitoring_context.logger.debug("Computing average per metric")
+        metrics_and_result: list[
+            typing.Union[
+                mm_results.ModelMonitoringApplicationMetric,
+                mm_results.ModelMonitoringApplicationResult,
+            ]
```

Comment:
`_get_metrics` returns only a list of `ModelMonitoringApplicationMetric`. therefore the typehint here should be:
`list[mm_results.ModelMonitoringApplicationMetric]`

Reply:
oh I see that in the next line you are adding mm_results.ModelMonitoringApplicationResult as well. Just a recommendation: maybe wrap these 2 operations in a different function `_add_metrics_and_general_drift` and then you can keep the same typehints: 
```
metrics_and_result: list[
            typing.Union[
                mm_results.ModelMonitoringApplicationMetric,
                mm_results.ModelMonitoringApplicationResult,
            ] = self._add_metrics_and_general_drift(metrics_per_feature)

```

Reply:
I think I will leave it like it is now, it is the only way the typing isn't yeling on me.